pkg/spdk/server.go

```diff
@@ -133,110 +134,117 @@ func (s *Server) tryEnsureSPDKTgtHealthy() error {
 
 	if running {
 		logrus.Info("SPDK Server: reconnecting to spdk_tgt")
-		return s.spdkClient.Reconnect()
+		return s.clientReconnect()
 	}
 
 	logrus.Info("SPDK Server: restarting spdk_tgt")
 	return util.StartSPDKTgtDaemon()
 }
 
+func (s *Server) clientReconnect() error {
+	s.Lock()
+	defer s.Unlock()
+
+	oldClient := s.spdkClient
+
+	client, err := spdkclient.NewClient(s.ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to create new SPDK client")
+	}
+	s.spdkClient = client
+
+	// Try the best effort to close the old client after a new client is created
+	err = oldClient.Close()
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to close old SPDK client")
+	}
+	return nil
+}
+
 func (s *Server) verify() (err error) {
 	replicaMap := map[string]*Replica{}
-	engineMap := map[string]*Engine{}
-	s.RLock()
+	replicaMapForSync := map[string]*Replica{}
+	engineMapForSync := map[string]*Engine{}
+
```

Comment:
What's the purpose of `replicaMapForSync`?

Reply:
This function copied 2 maps for the replicas. `replicaMap` will become the server variable `s.replicaMap`. `replicaMapForSync` will be used for each replica syncup `r.Sync(spdkClient)`.
Notice that when syncing up replicas, we won't use `s.Lock()` to protect the map iteration. We cannot use `replicaMap`( which is already `s.replicaMap`) for the iteration since other gRPC calls like `ReplicaCreate` can change the map.